src/ctap/storage.rs

```diff
@@ -467,6 +471,70 @@ impl PersistentStore {
         )?)
     }
 
+    /// Reads the byte vector stored as the serialized large blobs array.
+    ///
+    /// If too few bytes exist at that offset, return the maximum number
+    /// available. This includes cases of offset being beyond the stored array.
+    pub fn get_large_blob_array(
+        &self,
+        mut byte_count: usize,
+        mut offset: usize,
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        if self.store.find(key::LARGE_BLOB_SHARDS.start)?.is_none() {
+            return Ok(vec![
+                0x80, 0x76, 0xBE, 0x8B, 0x52, 0x8D, 0x00, 0x75, 0xF7, 0xAA, 0xE9, 0x8D, 0x6F, 0xA5,
+                0x7A, 0x6D, 0x3C,
+            ]);
+        }
+        let mut output = Vec::with_capacity(byte_count);
+        while byte_count > 0 {
+            let shard_offset = offset % SHARD_SIZE;
+            let shard_length = cmp::min(SHARD_SIZE - shard_offset, byte_count);
+
+            let shard_key = key::LARGE_BLOB_SHARDS.start + offset / SHARD_SIZE;
+            if !key::LARGE_BLOB_SHARDS.contains(&shard_key) {
+                // This request should have been caught at application level.
+                return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+            }
+            let shard_entry = self.store.find(shard_key)?.unwrap_or_default();
+            if shard_entry.len() < shard_offset + shard_length {
+                // If fewer bytes exist than requested, return them all.
+                output.extend(&shard_entry[shard_offset..]);
+                return Ok(output);
+            }
+            output.extend(&shard_entry[shard_offset..shard_offset + shard_length]);
+            offset += shard_length;
+            byte_count -= shard_length;
+        }
+        Ok(output)
+    }
+
+    /// Sets a byte vector as the serialized large blobs array.
+    pub fn commit_large_blob_array(
+        &mut self,
+        large_blob_array: &[u8],
+    ) -> Result<(), Ctap2StatusCode> {
+        if large_blob_array.len() > MAX_LARGE_BLOB_ARRAY_SIZE {
+            return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+        }
+        const MIN_SHARD_KEY: usize = key::LARGE_BLOB_SHARDS.start;
+        const SHARD_COUNT: usize = key::LARGE_BLOB_SHARDS.end - MIN_SHARD_KEY;
+        let mut transactions = Vec::with_capacity(SHARD_COUNT);
+        for shard_key in MIN_SHARD_KEY..key::LARGE_BLOB_SHARDS.end {
+            let large_blob_index = (shard_key - MIN_SHARD_KEY) * SHARD_SIZE;
+            if large_blob_array.len() > large_blob_index {
+                let shard_length = cmp::min(SHARD_SIZE, large_blob_array.len() - large_blob_index);
+                transactions.push(StoreUpdate::Insert {
+                    key: shard_key,
+                    value: &large_blob_array[large_blob_index..large_blob_index + shard_length],
+                });
+            } else {
+                transactions.push(StoreUpdate::Remove { key: shard_key });
+            }
+        }
```

Comment:
To avoid arithmetic, this loop could be more simply written as:

```rust
let mut shards = large_blob_array.chunks(SHARD_SIZE);
let updates = key::LARGE_BLOB_SHARDS.map(|key| match shards.next() {
    None => StoreUpdate::Remove { key },
    Some(value) => StoreUpdate::Insert { key, value },
}).collect();
```

Reply:
`chunks` is super useful here! I don't think I can collect as nice with the improvement mentioned below to skip some `remove`s. Maybe with a filter_map, but the break seems more natural to me. Also FYI, your code give me a:

> error[E0277]: the size for values of type `[persistent_store::store::StoreUpdate<_>]` cannot be known at compilation time

Reply:
Obsolete.